src/hotspot/cpu/x86/assembler_x86.cpp

```diff
@@ -13364,6 +13364,16 @@ void Assembler::mov64(Register dst, int64_t imm64, relocInfo::relocType rtype, i
   emit_data64(imm64, rtype, format);
 }
 
+#ifdef _LP64
+void Assembler::mov32_or_64(Register dst, int64_t imm) {
+  if ((uint64_t)imm < nth_bit(32)) {
```

Comment:
Drive-by comments:
 a) macro-assembler stuff like this should be in macroAssembler;
 b) there is `is_simm32(imm)` for checks like these;
 c) I did [JDK-8319406](https://bugs.openjdk.org/browse/JDK-8319406) recently, maybe you could just use that?

Reply:
I also wonder if `movl` is actually correct at 64-bit, since it would likely only update the lower 32 bits of the register. [JDK-8319406](https://bugs.openjdk.org/browse/JDK-8319406) does it right with `movq r/m64, imm32`.

Reply:
God you are fast. 

movptr looks suitable. Why did you name it "ptr" if its not for pointers?

Reply:
I haven't changed the name of the method. `movptr` means "move something of the width of the machine pointer", like everywhere else in assembler code. That fits your case, I think?

Reply:
Hmm, movptr only works its magic if the input is smaller than signed int max (2gb), and it needs one byte more than my variant.

base < 2g:

```
   35  ;; decode_klass_not_null
   36   0x00007efdcc9e51c4:   mov    $0x27000000,%r11                                                                                                                                                                                                                               
   37   0x00007efdcc9e51cb:   add    %r11,%r10
```

base > 2g:

```
   35  ;; decode_klass_not_null
   36   0x00007ff7a89e51c4:   movabs $0x82000000,%r11
   37   0x00007ff7a89e51ce:   add    %r11,%r10          
```

Reply:
FYI the logic for immediate matching is:

    if (is_uimm32(imm)) {
        movl(dst, imm);
    } else if (is_simm32(imm)) {
        movq(dst, imm);
    } else {
        mov64(dst, imm);
    }

The reason is that `movl` is smaller than `movq` in code size.

I hope that I do not miss anything here.

Reply:
Yeah, current `movptr` is sign-extending 32->64, which is not extra-efficient for 32-bit unsigned imms. I think we can indeed check for `is_uimm32` in `movptr` to cover that case too.

Reply:
This seems to work:

```
diff --git a/src/hotspot/cpu/x86/macroAssembler_x86.cpp b/src/hotspot/cpu/x86/macroAssembler_x86.cpp
index 88296656485..ba4b089c7aa 100644
--- a/src/hotspot/cpu/x86/macroAssembler_x86.cpp
+++ b/src/hotspot/cpu/x86/macroAssembler_x86.cpp
@@ -2568,7 +2568,9 @@ void MacroAssembler::movptr(Register dst, Address src) {
 // src should NEVER be a real pointer. Use AddressLiteral for true pointers
 void MacroAssembler::movptr(Register dst, intptr_t src) {
 #ifdef _LP64
-  if (is_simm32(src)) {
+  if (is_uimm32(src)) {
+    movl(dst, checked_cast<uint32_t>(src));
+  } else if (is_simm32(src)) {
     movq(dst, checked_cast<int32_t>(src));
   } else {
     mov64(dst, src);
diff --git a/src/hotspot/share/asm/assembler.hpp b/src/hotspot/share/asm/assembler.hpp
index 7b7dbd4ede7..a533b963844 100644
--- a/src/hotspot/share/asm/assembler.hpp
+++ b/src/hotspot/share/asm/assembler.hpp
@@ -359,6 +359,7 @@ class AbstractAssembler : public ResourceObj  {
   }

   static bool is_uimm12(uint64_t x) { return is_uimm(x, 12); }
+  static bool is_uimm32(uint64_t x) { return is_uimm(x, 32); }

   // Accessors
   CodeSection*  code_section() const   { return _code_section; }
```

Reply:
That change seems to hold water on its own, saving quite a bit of code. Filed: [JDK-8323503](https://bugs.openjdk.org/browse/JDK-8323503).

Reply:
@shipilev I like this. I'll wait for your patch to go in.

Reply:
I think it would be better to move to `movptr` here and start testing it. I suspect there are a few places in Hotspot where we count the actual instruction _size_ for some stuff, and even the `mov64` -> `movptr` rewrite without `movl` optimization would highlight it. I agree the final version should be tested when JDK-8323503 is in.

Reply:
> FYI the logic for immediate matching is:

@merykitty, could you point me where that logic is? Cannot find it in current Hotspot sources, and would like to reference it in my PR.

Reply:
@shipilev It is just the code representation of the matching, the corresponding nodes are `loadConUL32`, `loadConL32` and `loadConL` and the matcher uses node costs to sort out the priority.

https://github.com/openjdk/jdk/blob/c1282b57f50002edd08c93aed784390cca83b9b8/src/hotspot/cpu/x86/x86_64.ad#L4807

Reply:
Oh, subtle. Thanks!

Reply:
With the new movptr variant, things look good:

< 2g

```
 ;; decode_klass_not_null
  0x00007fcb608681c4:   mov    $0x18000000,%r11d
0x00007fcb608681c4:   41 bb 00 00 00 18 
  0x00007fcb608681ca:   add    %r11,%r10
0x00007fcb608681ca:   4d 03 d3 
```

< 4g

```
 ;; decode_klass_not_null
  0x00007f382c8681c4:   mov    $0x85000000,%r11d
0x00007f382c8681c4:   41 bb 00 00 00 85 
  0x00007f382c8681ca:   add    %r11,%r10
0x00007f382c8681ca:   4d 03 d3 
```

> 4g

```
 ;; decode_klass_not_null
  0x00007fd8908681c4:   movabs $0x7fd7b6000000,%r11
0x00007fd8908681c4:   49 bb 00 00 00 b6 d7 7f 00 00 
  0x00007fd8908681ce:   add    %r11,%r10
0x00007fd8908681ce:   4d 03 d3 
```

I assume if the 32-bit mov variants were to use non-Rxx registers, the REX prefix could also be omitted? Giving us another byte back.

I'll run tests with the new variant.

Reply:
BTW, I think it would be useful for PrintAssembly to also optionally print the raw bytes in addition to the decoded instruction. It certainly helps me understanding stuff. Do you agree? If yes, I'll prepare a patch.

Reply:
I'm pretty marginal on this. If I ever have needed to see bytes, I've used the debugger.